### ISSUE TITLE
Author merge fails

### ISSUE THREAD

**Issue by tfmorris** [internetarchive/openlibrary#5594]
I know there was an issue about failing author merges, but I can't find it. 
### Evidence / Screenshot (if possible)

![Screen Shot 2021-08-30 at 12 06 24 PM](https://user-images.githubusercontent.com/82178/131369876-172fcff4-a413-46f1-845d-5ac19755219d.png)

### Relevant url?
https://openlibrary.org/authors/OL6736985A/Lucy_Leffingwell_Cable_Bikl%C3%A9?merge=true&duplicates=OL2225567A,OL5767240A,OL6636018A,OL3665163A

### Steps to Reproduce

1. Go to the url
2. Do click it

* Actual: Error saying "That merge didn't work. It's our fault, and we've made a note of it."
* Expected: Merged author record

### Details

- **Logged in (Y/N)?** 
- **Browser type/version?** 
- **Operating system?** 
- **Environment (prod/dev/local)?** prod

### Proposal & Constraints

In addition to fixing the bug, I'm wondering if the error message is accurate. I've received hundreds (thousands?) of these and I've never seen any action taken as a result of this "note taking." Does someone actually review these and fix them? If not, perhaps the messaging should be changed to something more accurate.

**Comment by seabelis** [comment]
- [ ] https://openlibrary.org/authors/OL9361307A/Edgar_Allan_Poe_%28duplicate%29

**Comment by seabelis** [comment]
- [ ] https://openlibrary.org/authors/merge?key=OL26680A&key=OL3014177A&key=OL7104908A

**Comment by dcapillae** [comment]
- [ ] https://openlibrary.org/authors/merge?key=OL215610A&key=OL7349088A

**Comment by LeadSongDog** [comment]
Some large merges now work that previously did not. Danielle Steel just did. Then, some are direction-particular: 
https://openlibrary.org/authors/merge?key=OL235327A&key=OL6890242A 
failed but not 
https://openlibrary.org/authors/merge?key=OL6890242A&key=OL235327A

**Comment by seabelis** [comment]
- [ ] https://openlibrary.org/authors/merge?key=OL934457A&key=OL2762549A

**Comment by tmanarl** [comment]
https://openlibrary.org/authors/merge?key=OL656172A&key=OL7552034A

**Comment by LeadSongDog** [comment]
The reverse merge seems to work on testing: https://testing.openlibrary.org/authors/OL656172A/James_Oliver_Curwood?_compare=Compare&b=11&a=10&m=diff

the complete record is at https://testing.openlibrary.org/authors/OL7552034A/James_Oliver_Curwood

while no works show at https://testing.openlibrary.org/authors/OL656172A/James_Oliver_Curwood

**Comment by tmanarl** [comment]
https://openlibrary.org/authors/merge?key=OL2632751A&key=OL2195639A

**Comment by tmanarl** [comment]
https://openlibrary.org/authors/merge?key=OL446457A&key=OL2763758A

**Comment by tmanarl** [comment]
https://openlibrary.org/authors/merge?key=OL28257A&key=OL5949173A

**Comment by LeadSongDog** [comment]
That last merge will likely work now after this: https://openlibrary.org/works/OL46876W/The_Lost_World?_compare=Compare&b=55&a=49&m=diff

That work record had both author records linked, which seems a common factor in these. It somehow took several edit attempts to change that double linkage.

**Comment by tmanarl** [comment]
https://openlibrary.org/authors/merge?key=OL6835818A&key=OL6766129A

**Comment by LeadSongDog** [comment]
@tmanarl That last is another that worked in the other direction.

**Comment by tmanarl** [comment]
What do you mean by reverse merge?

Travis Pollan

**Comment by LeadSongDog** [comment]
Merges default to merging into the most-populated author as primary. Sometimes switching to the least-populated as primary will work when the default fails, as for: 

![D19F19EB-962C-44B1-AB70-9EA990408240](https://user-images.githubusercontent.com/11435431/150909269-dbcf8e8d-793f-41f1-a7b2-518aae108478.jpeg)

**Comment by dcapillae** [comment]
@LeadSongDog  It is not possible choose the main record arbitrarily according to the librarians' guidelines. The main record is always the oldest one.

**Comment by tmanarl** [comment]
My thoughts exactly. Merging an older record into a newer one defeats the purpose of trying to keep the oldest as main. Those errors have been the ones I’ve been posting here.

Travis Pollan

**Comment by LeadSongDog** [comment]
Possible but not desirable. I’m not suggesting it is good practice, rather noting that the choice of primary may have a bearing on diagnosing the failure. I strongly suspect the root problem is that one or mor editions have direct linkage to the  wrong author record, a situation which can’t be corrected by regular users.

**Comment by tfmorris** [comment]
This is off the topic of this issue, but

> It is not possible choose the main record arbitrarily according to the librarians' guidelines. The main record is always the oldest one.

@dcapillae If that's the guideline, which seems like a reasonable one for **most** cases, the software should support/enforce it. Human nature is to accept the system provided default and the system chooses the author record with the most works, not the oldest. The only way one can find the old record is by hovering over the links and squinting at the author ID to try and discern whether 500000 is bigger or smaller than 600000. 

And, as evidenced by this thread, when the system fails to work at all, users will resort to creative workarounds in an attempt to accomplish their task.

This issue has been accumulating lots of evidence that the problem is real, but no one is bothering to actually **_fix_** it.

**Comment by tmanarl** [comment]
I open each author record in a new tab, scroll to the bottom to see which one has the oldest creation date and then attempt to merge into the oldest record. Those merges that fail I post here. You are correct in that the top entry is not always the oldest record. 

Travis Pollan

**Comment by tfmorris** [comment]
> I open each author record in a new tab, scroll to the bottom to see which one has the oldest creation date and then attempt to merge into the oldest record. 

It's unreasonable to expect ordinary users to invest this level of **extra** effort when a) they're already volunteering their time and b) it's trivial for the computer to do.

**Comment by mekarpeles** [comment]
I agree w/ @tfmorris, at least by default it seems like having the platform try to select the oldest record seems like it could save time. I wonder if it may deter double-checking though/then.

**Comment by mekarpeles** [comment]
@seabelis I'm adding you as the lead if that's okay -- there seem to be a handful of failures listed here and I guess I wasn't fully sensitive to the scope. If this is something engineering can help with, please let us know how :)

**Comment by tmanarl** [comment]
https://openlibrary.org/authors/merge?key=OL152307A&key=OL7889785A

**Comment by cdrini** [comment]
Oh, this is definitely an engineering issue. This is causing a lot more friction to our librarians than I realized.

**Comment by cdrini** [comment]
This example is I think the most useful one for debugging:

https://openlibrary.org/authors/merge?key=OL3745151A&key=OL9158169A

The merge request will try to update all these editions: http://openlibrary.org/query.json?type=/type/edition&authors=/authors/OL9158169A&* by replacing OL9158169A with OL3745151A

If we do a copydocs of all those editions/works into the local environment (i.e. `./copydocs.py /authors/OL9158169A /authors/OL3745151A`, and then copy those editions/works `./copydocs.py $(curl -s 'http://openlibrary.org/query.json?type=/type/edition&authors=/authors/OL9158169A&key' | grep -Eo "\/books\/OL[0-9]+M")` Then locally go to the merge page `/authors/merge?key=OL3745151A&key=OL9158169A` (after logging in as admin) to try to debug what's happening.

- You'll need to use the debugger: https://github.com/internetarchive/openlibrary/wiki/Debugging#using-a-debugger
- I haven't tested the above copy docs queries, so might need some massaging. See https://github.com/internetarchive/openlibrary/wiki/Getting-Started#importing-test-data

**Comment by mheiman** [comment]
I've been digging into this one a little bit, but I can't reproduce the problem in development.

I've taken several of the merges listed above, verified that they're still failing, and copied all of the data to a development instance as described by @cdrini above.

All of the merges work in development. I'm guessing there's something subtle in the database that's lost when author and book records are exported and imported.

Are there any tools for actually replicating chunks of the database to a development server?

**Comment by seabelis** [comment]
- [ ] https://openlibrary.org/authors/OL2657659A/Ambrose_Bierce_%28duplicate%29

**Comment by seabelis** [comment]
Regarding the oldest being selected by default, I'm on board with that. This is how the works merge interface works. The authors are presented with the most relevant/most populated first because that's how search works. When this interface was created back in 2010 or so, the most populated were typically the oldest. Now this is no longer the case. If the easy fix is to display the created date on the merge form, I think that would be a step in the right direction.

**Comment by LeadSongDog** [comment]
I hate to keep harping on this, but just look at that example. Comparing the author IDs in 
https://openlibrary.org/books/OL31884994M.json 
(OL7202807A and OL5251200A)
to those in the linked work 
https://openlibrary.org/works/OL24159503W.json 
(OL3745151A and OL5251200A)
there is still a clear problem: editions should not have explicit links to author redirects.

**Comment by mekarpeles** [comment]
@cdrini maybe as part of our resolving redirects work we also need to somehow resolve authors or at least handle this case under the hood?

**Comment by LeadSongDog** [comment]
Handling under the hood is likely to mask rather than fix problems, especially when one or more of the editions were linked to undifferentiated author records. The worst part of the problem is that edition authors are not exposed in the UI (except as json) nor are they correctable by unprivileged users. If an edition is linked to a different John Smith than the work, that linkage May never be noticed let alone fixed.

**Comment by mheiman** [comment]
@cdrini and I spent some time trying to understand the reproducibility problem. No resolution was reached, but we found a few things that _aren't_ causing the merge failure.

1. Importing records to a dev instance with copydocs.py can result in small changes to the author YAML data structure (particularly to the 'bio' field). It would appear the the author data definition changed at some point, but older author records on production that haven't been updated recently still have the old structure.  _However,_ updating the author data on production to apply the new structure doesn't fix the merge problem.

2. In some of the failure cases listed above, the secondary author (the one being merged into the main author) has associated works, but those works aren't indexed by SOLR, so they don't show up in the merge interface or on the author page. They can be made to appear by explicitly requesting a SOLR update for the affected works. _However,_ doing so doesn't fix the merge problem.

We'll continue trying to figure out why the problem goes away when the data is moved to another instance.

**Comment by mheiman** [comment]
I believe I've tracked down the problem, though I'm not sure what the solution is. @LeadSongDog was on the right track.

### Detailed Explanation:
In digging deeper to figure out why I couldn't reproduce the problem in development, I discovered that when edition records are moved using copydocs.py, any author values in the record are stripped out. This means that none of the editions are pulled up when the merge process is looking for records to update. On production, on the other hand, those edition records _are_ included in the merge process, because the author data is intact.

Working with the example merge @cdrini cites above (https://openlibrary.org/authors/merge?key=OL3745151A&key=OL9158169A) I copied all the records as described. Then, in order to make sure my dev data was actually equivalent, I manually downloaded the 10 edition records in YAML format, and used the YAML editing interface to update my dev records to match the production data.

In that process, I discovered that there was one edition record that could not be saved, generating the useful error "Unable to connect to infobase server" (this is a lie). I suspect if someone with appropriate privs on production attempts to save this record, the same thing will happen:

https://openlibrary.org/books/OL33913405M.yml?m=edit

I traced the actual error to /openlibrary/infogami/infobase/client.py where RemoteConnection is receiving this message from InfoBase:

"expected /type/author, found /type/redirect", "at": {"key": "/books/OL33913405M", "property": "authors"}, "value": "/authors/OL9547707A"}

/authors/OL9547707A resolves to a redirect, and the save process is choking because it expects an author. Removing that entry from the YAML file allows the record to be saved.

### Bottom line:
When the author merge process runs, it searches for all records containing the old author ID, updates just that field, and passes all the updated records in a batch to the save_many endpoint. If any of those records contain an author redirect, the save process dies ungracefully and the merge fails.

### Possible resolutions:
I'm still learning the codebase, so I don't know what's most practical here.

- Modify the author merge process to find and fix author redirects in edition records before passing them to InfoBase
- Teach InfoBase how to resolve author redirects in records that it's trying to save
- Batch fix all author redirects in edition records
- Remove author data from edition records entirely (is this a legacy thing?)

**Comment by cdrini** [comment]
Great sleuthing, @mheiman ! That seems like the ticket! That's a bit of a disturbing error 😬 

I've suspected there's been a bug in this code for a while (this is the code that finds the records to update when an author merge is kicked off):

https://github.com/internetarchive/openlibrary/blob/39962157c12465d050e43458444aba9a59661c8d/openlibrary/plugins/upstream/merge_authors.py#L132-L141

The `type/edition` query has a different format for `authors` than the `type/work` query; I think it might never be matching anything?

I'm _super_ surprised it won't let you save an edit when one of the fields is a redirect.....

I tried editing an unrelated field on that edition in prod, and it saved correctly. Buuut, if I try to edit the author field in any way, it errors. So it must be applying an extra level of validation when the authors field is touched, causing it to error.

So then it probably _is_ updating the authors field, so the query must be matching at least sometimes :/ 

So I think in terms of your resolutions, 1 and 3 seem :+1: ; 2 I think would be difficult, and potentially a little too magical (we generally like to avoid having "secret" edits happen when a user does something, because it makes things difficult to debug and track -- like the `type/text` charade we ran through :P). 4 doesn't really have consensus; there's a decent amount of discussion around it; see #5594 .

**Comment by mheiman** [comment]
Thanks, @cdrini. If you think that adding some additional cleanup code to the merge routine is the best solution, I'm happy to keep working on this. It's going to be tricky to test, since the editing tools won't let me create bad records, but maybe I can poke the values directly into the database.

**Comment by cdrini** [comment]
Hehe maybe! Good luck :P I think investigating that snippet might help prevent new occurrences of this issue appearing (1). For 1 specifically we'd want to confirm that editions in the database are indeed getting updated when authors are merged. If they were, then there would be no errors on these merges.

Doing a bulk update would allow all the past stuck merges to go through (3). That's probably best done as a one-off script to run in a bot-enabled account.

**Comment by mheiman** [comment]
Editions in the database are definitely getting updated during an author merge — I confirmed that in my testing. The snippet you cited appears to be working as intended. Things just go wrong when those edition records also contain author redirects.

We're going to need something that scans through all the edition records involved in the merge before they're saved, makes sure the author entries resolve, and replaces any redirects with their resolved key. That will allow all the current stuck merges to go through, and prevent future failures from this source.

I'll see what I can come up with...

**Comment by cdrini** [comment]
We're not sure what caused the record to be not updated. Here is the merge that caused OL33913405M to become inconsistent, and have a reference to an author redirect: https://openlibrary.org/recentchanges/2021/10/08/merge-authors/87081389

**Comment by cdrini** [comment]
Here is the dump of the edition records that need to be fixed (55k): [sql.csv.gz](https://github.com/internetarchive/openlibrary/files/8421638/sql.csv.gz)

Here's the format, as a csv:

<html><body>


Edition | LastModified | Author | RedirectLocation | RedirectLastModified
-- | -- | -- | -- | --
/books/OL15350969M | 2009-12-15 02:41:02.077848 | /authors/OL4624437A | /authors/OL184870A | 2011-06-10 01:01:01.009647
/books/OL20643740M | 2008-10-29 07:17:22.032719 | /authors/OL2542780A | /authors/OL74978A | 2009-05-14 09:20:58.783037
/books/OL18664950M | 2008-10-18 23:49:42.418645 | /authors/OL2542780A | /authors/OL74978A | 2009-05-14 09:20:58.783037
/books/OL21852313M | 2008-11-04 10:17:39.796926 | /authors/OL4280812A | /authors/OL486177A | 2009-05-14 09:16:21.763386
/books/OL22913513M | 2009-02-12 02:23:34.26629 | /authors/OL2542780A | /authors/OL74978A | 2009-05-14 09:20:58.783037
/books/OL22913508M | 2009-02-12 02:23:30.740693 | /authors/OL2542780A | /authors/OL74978A | 2009-05-14 09:20:58.783037


</body>
</html>

**Comment by mheiman** [comment]
We've updated more than half a million edition records that contained author redirects. This should resolve all of the author merge failures previously reported -- if you reported a failure earlier in this thread, please try to push it through now.

We still don't know exactly how the bad records are being created, so I wouldn't consider this issue closed until we solve that. I have a hunch that I'm pursuing.

If you encounter any more author merges that fail, please continue to report them here — that will give us more data to track down the source of the bad data.

**Comment by dcapillae** [comment]
I have merged the authors included in my [list of merge failures](https://github.com/internetarchive/openlibrary/issues/5594#issuecomment-980444196) except for two authors where it continues to fail:

- [ ] Christopher Paolini: https://openlibrary.org/authors/merge?key=OL1425963A&key=OL7929134A
- [ ] Stephen Crane: https://openlibrary.org/authors/merge?key=OL24527A&key=OL7391065A

**Comment by LeadSongDog** [comment]
Looking at https://openlibrary.org/works/OL5819884W, one sees both author records on one work. Further, individual editions show links to either author record, not just one: OL35665841M links to OL7929134A, while OL27301994M links to OL1425963A. It would be instructive to either choose or just remove these redundant links from the edition records: there’s no need for editions to show a redirected author record linkage.

**Comment by mheiman** [comment]
@dcapillae, I tracked down the problem editions and pushed those two merges through, so you can finish your cleanup work on those authors.

**Comment by dcapillae** [comment]
I get a new merge failure. I do not know if it is related:

[] Avi: https://openlibrary.org/authors/merge?key=OL27173A&key=OL9250111A

**Comment by mheiman** [comment]
There are still some author redirects out there that our cleanup attempts missed. It looks like there was a big batch of Robert Louis Stevenson that had bad data, and that was affecting anyone who contributed to a Stevenson adaptation, like Avi.  I did a cleanup run and was able to push this merge through.

Please keep reporting problems. Thanks!

**Comment by tfmorris** [comment]
The behavior has changed so that merges no fail silently instead of giving an error. :(  It looks like the server is returning a 400 which isn't getting reported to the user as an error.

The Edgar Allen Poe merge reported by @seabelis still fails https://openlibrary.org/authors/OL28127A/Edgar_Allan_Poe?merge=true&duplicates=OL9361307A
https://openlibrary.org/authors/OL9361307A/Edgar_Allan_Poe_%28duplicate%29

Also Goethe (not sure which record is the culprit) https://openlibrary.org/authors/OL13193A/Johann_Wolfgang_von_Goethe?merge=true&duplicates=OL9962258A,OL9966231A,OL10034465A,OL10349418A,OL11432417A,OL10139257A,OL11039996A

Arthur Conan Doyle https://openlibrary.org/authors/OL471614A/Arthur_Conan_Doyle?merge=true&duplicates=OL10715195A,OL7630022A,OL7514077A,OL11432680A,OL11442105A,OL11436184A,OL10791911A,OL10792985A,OL10834882A,OL10718095A,OL10717518A,OL10757749A,OL11412851A,OL11501545A,OL11457176A,OL11019600A,OL11091013A,OL11110083A,OL11151515A,OL10860030A,OL10867145A,OL11514358A,OL11521018A

@mheiman Perhaps I'm being naive, but if the condition that causes the problem is known, why not fix the code to be resilient to it, rather than attempting multiple rounds of database cleanup?

**Comment by mheiman** [comment]
It's not an either/or situation. I've done some digging into the underlying problem, but it requires some deep changes in Infogami, so it's going to take some careful work to get a solid solution, and looks like it may require touching a lot of different parts of the code. It's on my list once I finish the project I'm currently working on. Until then I've been fixing specific cases as I find them.

**Comment by tfmorris** [comment]
Here's a recent one which fails from #8042:
https://openlibrary.org/authors/OL16963A/Emile_Monte%CC%81gut?merge=true&duplicates=OL1964447A,OL3450354A,OL4279219A,OL4404792A,OL6023387A,OL6056363A,OL6184875A,OL6389150A
and it looks like the ones in my Jan. 2023 comment still fail too https://github.com/internetarchive/openlibrary/issues/5594#issuecomment-1370221173

**Comment by LeadSongDog** [comment]
@tfmorris As is frequently the case, the oldest record there was not the best (name was wrong and external IDs were omitted). Also, https://openlibrary.org/authors/OL6056363A lists several works that show problematic co-authors that will need manual curation.  Merging the other seven author records should, I think, succeed.

**Comment by tfmorris** [comment]
The situation has improved slightly in that the error response now includes the reason for the failure (e.g. OLID which is a redirect and the work/edition that contains it), which makes it possible to repair the data. Unfortunately, this information isn't displayed to the user, so you have to know how to find it from the developer console of your browser.
